include: "rules/common.smk"

##### Target Rules #####

# define rules that don't need to be run on a compute node.
# i.e. those that can be run locally.
localrules: all, echo_RG_fixed, genome_faidx, genome_dict, make_scaff_group_interval_lists, make_chromo_interval_lists, make_scatter_interval_lists,
# other local rules
# combine_bcftools_stats, gatk3_register, species_sample_lists, get_genome_length, get_ave_depths


rule all:
    input:
        expand("results/hpsmc/split-time-sim/greenback---x---s_hayden_*.psmc"),
        #expand("results/inbreeding/het/{s}-het-perc.txt", s=sample_list),
        #"results/inbreeding/het/combined-het-percents.tsv",
        #expand("results/hpsmc/hpsmc-plot/{p1}---x---{p2}", p1=pop1, p2=pop2),
        
        #"results/plink/pw-fst/co-lineages-aut-snps-0.05-fst",
        #"results/plink/phylip/co-lineages-aut-snps-0.05",
        #"results/plink/pca/co-lineages-aut-snps-0.05-pruned-pca"
        #"results/plink/ld-prune/co-lineages-aut-snps-0.05",
        #"results/plink/allele-freq/co-lineages-aut-snps-0.05",
        #expand("results/bcf/{b}-aut-bisnps-maf-0.05.bcf", b=bcfsubsamp),
        #expand("results/tree/aut-snps-{maf}-tree", maf=mafs),
        #expand("results/plink/het/aut-snps-{maf}", maf=mafs),
        #expand("results/plink/het/aut-snps-{maf}-pruned", maf=mafs),
        #expand("results/plink/pw-fst/aut-snps-{maf}-fst", maf=mafs),
        #expand("results/plink/phylip/aut-snps-{maf}", maf=mafs),
        #expand("results/plink/allele-freq/pop-counts/aut-snps-{maf}", maf=mafs),
        #expand("results/plink/pca/aut-snps-{maf}-ld-pruned-pca", maf=mafs),
        #expand("results/hpsmc-test/haploidize_bam_sect/{hp}/{hpc}_haploidized.fa", hp=hpsmcpops, hpc=hpsmcchroms),
        #expand("results/hpsmc/hpsmc-plot/{p1}---x---{p2}", p1=pop1, p2=pop2),
        #expand("results/hpsmc/haploidized_bam/{hp}_haploidized.fa", hp=hpsmcpops),
        #expand("results/hpsmc-test/psmcfa-from-2-fastas/{p1}---x---{p2}.psmcfa", p1=pop1, p2=pop2),
        #expand("results/plink/pca/aut-snps-{maf}-pca", maf=mafs),
        #misc in progress outputs
        #expand("results/bcf/all.bcf"), #hardfiltering final outputs
        #expand("results/bcf/all.bcf.csi"), #hardfiltering final outputs
        #expand("results/roh/allele-freq/snps-maf-0.05/{pop}-freqs.tabs.gz", pop=population),
        #expand("results/roh/snps-maf-0.05/{pop}-roh.txt", pop=population)
        #expand("results/psmc/psmc-plot/{subsamp}/{subsamp}", subsamp=subsamp),
        #expand("results/psmc/psmc-plot/{subsamp}/{subsamp}", subsamp=subsamp)
        #expand("results/psmc/run-psmc/{s}.psmc", s=sample_list),
        #"results/plink/pw-fst-all/fst-all", 
        #"results/plink/pw-fst-snps-0.05/fst-snps-0.05"
        
        #expand("results/psmc/remove-sex-bams/aut_{s}.bam", s=sample_list),
        #expand("results/psmc/psmc-consensus-sequence/{s}.fq.gz", s=sample_list)
        #expand("results/psmc/psmcfa/{s}.psmcfa", s=sample_list)
        #expand("results/psmc/run-psmc/{s}.psmc", s=sample_list),
        #expand("results/psmc/psmc-plot/{subsamp}/{subsamp}", subsamp=subsamp),

        #plink-pca outputs
        #"results/bcf/filt_biallelic_maf_0.05/main.bcf"
        #"results/plink/pca/snps-no-y-pca"
        #"results/plink/allele-freq/snps-no-y"   
        
        #hpsmc outputs
        #expand("results/hpsmc/haploidize_bam_sect/{s}/{chrom}_haploidized.fa", s=sample_list, chrom=unique_chromosomes)
        #expand("results/hpsmc/psmcfa-from-2-fastas/{p1}---x---{p2}.psmcfa", p1=pop1, p2=pop2),
        #expand("results/hpsmc/run-hpsmc/{p1}---x---{p2}.psmc", p1=pop1, p2=pop2),
        #expand("results/hpsmc/hpsmc-plot/{p1}---x---{p2}", p1=pop1, p2=pop2),

        #final outputs
        #expand("results/mapping/mkdup/mkdup-{s}.bam", s=sample_list), #mapping final output
        #expand("results/mapping/gatk-rmdup/{s}.bam", s=sample_list), #mapping rmdup final output
        #expand("results/angsd_bams/overlap_clipped/{s}.bam", s=sample_list), #angsd-bams output
        #expand("results/calling/corrected_missing_vcf_sect/{sgc}.vcf.gz", sgc=sg_or_chrom), #calling final output
        #expand("results/bcf/all.bcf"), #hardfiltering final outputs
        #expand("results/bcf/all.bcf.csi"), #hardfiltering final outputs
        
        #troubleshooting RG errors
        #expand("results/angsd_bams/echo-RG-fixed/{s}.txt", s=sample_list),
        
        #calling intermediate outputs 
        #expand("results/calling/interval_lists/{scaff_group}.list", scaff_group=unique_scaff_groups),
        #expand("results/calling/interval_lists/{chromo}.list", chromo=unique_chromosomes),
        #expand("results/calling/vcf_sections/{sgc}/{scat}.vcf.gz", sgc=sg_or_chrom, scat=unique_scats),
        #expand("results/calling/vcf_sections/{sgc}.vcf.gz", sgc=sg_or_chrom),
        #expand("results/calling/genomics_db/{sgc}", sgc=sg_or_chrom), #calling intermediate output
        #expand("results/calling/gvcf/{s}.{ext}", s=sample, ext = ["g.vcf.gz", "g.vcf.gz.tbi"]), #calling intermediate output
        #expand("results/vcf_sections/{sgc}.{ext}", sgc=sg_or_chrom, ext = ["vcf.gz", "vcf.gz.tbi"]), #calling intermediate output
        #expand("results/scatter_config/scatters_{bs}.tsv", bs=binsize), #generate scatters file from scatter_intervals_file
        
        #PSMC outputs
        #expand("results/psmc/split-sex-bams/aut_{s}.bam", s=sample_list),
        #expand("results/psmc/psmc-plot/{s}.eps", s=sample_list), #PSMC final outputs, does weird things with file endings to this needs some help
        #expand("results/psmc/psmc-plot/{s}.par", s=sample_list),
        #"results/psmc/psmc-plot/all/all-together",
        #expand("results/psmc-plot/by-subsamp/{psmc_id}/{subsamp}", subsamp=subsamp)

        #PCANGSD PCA outputs
        #expand("results/pca/scat_regions/{scat}.scat_regions.tsv", scat=unique_scats),
        #expand("results/bcf/beagle-gl/sections/{scat}.beagle-gl.gz", scat=unique_scats),
        #"results/bcf/beagle-gl/beagle-gl.gz",

        #qc outputs
        #expand("results/qc/samtools_stats/{s}.txt", s=sample_list),
        #expand("results/qc/bcftools_stats/all-{fc}.txt", fc=["ALL", "PASS", "FAIL"],),
        #expand("results/qc/bcftools_stats/all-pass-maf-{maf}.txt", maf=mafs,),
        #expand("results/qc/bcftools_stats/all-pass-maf-{maf}.txt", maf=config["bqsr_maf"], bq=[str(x) for x in range(0, int(config["bqsr_rounds"])+1)]),
        #expand("results/qc/coverage-depth/{s}.txt", s=sample_list),
        #"results/qc/coverage-depth/avg-count/above-10.txt",
        #"results/qc/coverage-depth/avg-count/avg-depth.txt",

        #installing flagfiles 
        #"results/flags/chromcompare_installed",
        #"results/flags/pcangsd_installed",

        #expand("results/qc/coverage/{s}.txt", s=SAMPLES),
        #"annotated/all.vcf.gz",
        #"qc/multiqc.html",
        #"plots/depths.svg",
        #"plots/allele-freqs.svg",
        


##### Modules #####

include: "rules/ref.smk"
include: "rules/mapping.smk"
include: "rules/angsd-ready-bams.smk"
include: "rules/calling.smk"
include: "rules/hardfilter.smk"
include: "rules/psmc.smk"
include: "rules/hpsmc.smk"
include: "rules/qc.smk"
include: "rules/destination-rules.smk"
include: "rules/plink.smk"
include: "rules/inbreeding.smk"
include: "rules/tree.smk"
#include: "rules/archive.smk"
#include: "rules/annotation.smk"
#include: "rules/downsample.smk"
#include: "rules/force-call-vcf.smk"